### ISSUE TITLE
`/eth/v1/validator/attestation_data` taking a long time to respond

### ISSUE THREAD

**Issue by mcdee** [Consensys-Incorporated/teku#7814]
### Description

Teku takes significantly longer to respond to a request for attestation data made on `/eth/v1/validator/attestation_data` than other beacon nodes.

A recent example on mainnet showing the time taken to receive attestation data from four different beacon nodes:

![screenshot_2023-12-11-234627](https://github.com/Consensys/teku/assets/511384/84ac1ba7-fc40-4690-8694-2f7a22f9d7f6)

The calls to lighthouse, nimbus and prysm took a few milliseconds, whereas teku took 1-2 orders of magnitude longer.  Given that the attestation data should come directly from head state, this seems like a long time to have to wait for the data.

**Frequency:** [How regularly does it occur?]

Every slot.

### Versions (Add all that apply)

teku/v23.10.0/linux-x86_64/-privatebuild-openjdk64bitservervm-java-17

**Comment by rolfyone** [comment]
It's a non trivial process currently because we're making sure fork choice is correct etc, and getting a signed block and state also, but we can run some profiling etc and see if there's improvements.

Is there a specific time vouch is calling for attestation_data? presumably just some time after block arrival... It'd just be helpful if theres specific context that we should investigate.

**Comment by rolfyone** [comment]
I've been able to replicate this if ForkChoice hasn't run already on a block.
Basically if Im running this query and the block hasn't been imported yet, or fully integrated i guess, then it'll get of the order of 120-160ms to generate attestation data.
Once that if i call around 4 seconds once the block has completed (or after seeing the slot line with a block if watching logging), Im getting about 3-8ms...
I'd expect given that you're consistently seeing this every slot that you might be calling before the block has been imported so it's missing head?

**Comment by mcdee** [comment]
> I'd expect given that you're consistently seeing this every slot that you might be calling before the block has been imported so it's missing head?

That's certainly plausible, although there seems to be something missing in this timeline.  We listen to all active beacon nodes for a `head` event.  When we see one, we wait 200ms to allow import on other beacon nodes that may take longer to complete importing the block, and then kick off the attestation process.  From what you say, this should be enough time for the import to complete unless we're seeing large discrepancies on block delivery time.

**Comment by rolfyone** [comment]
Hrm ok, so it'll be doing FcU prior to producing the first attestation still. There's possibly some improvements that would help, will add this to the backlog to consider...

Re. timings as a workaround, could consider reducing the timeout on teku (if that's easy).

Basically, we run FcU on block import, and if that has just been done, then another FcU is redundant, so doing it twice in a row is needless overhead. 

If a block is missed, then the FcU won't have been done, and in that situation it's important that it has been executed correctly still.

Historically we automatically promoted the first block without FcU and that was problematic, but flowing on from that we've got a further tweak to attestation production where by we can clean up this FcU if it's already been executed, so i'll mark this as a bug stemming from the fact that we could have removed this when we changed block processing.
